Add CI for Pylero

ntkathole opened the pull request:
This PR includes 
- Github action CI for pull requests.
- Added pre-commit linting for pylero.
- Added note in readme about pre-commit.

PR raised from my fork, you can check CI running at https://github.com/ntkathole/pylero/pull/1 

![Screenshot_2021-06-05 Added CI for Pylero by ntkathole · Pull Request #1 · ntkathole pylero](https://user-images.githubusercontent.com/16951913/120883728-6f9e2e00-c5fc-11eb-831f-73654df438d6.png)

ntkathole committed:
Added CI for Pylero

ntkathole committed:
Added README note for pre-commit

leelavg commented on README.md:
```diff
@@ -260,3 +260,11 @@ req = Requirement(project_id="myproj", work_item_id="MYPROJ-11")
+## Before you commit
+
+In order to ensure you are able to pass the GitHub CI build, it is recommended that you run the following commands in the base of your pylero directory
+
```
Would be a good idea to add `$ pip install pre-commit` as well?

leelavg commented on .github/workflows/pull_request.yml:
```diff
@@ -0,0 +1,30 @@
+          pip install -U pip
+          pip install -U -r requirements.txt
+          pip install .
+      - name: Pre Commit Checks
```
Unfortunately, we do have a lot of users installing pylero with py2.7. Can you please include that in the matrix but exclude this step?

Rationale being, we want to check installation is going well even with py2.7 and at the same time don't want all precommit checks with it, sounds good?

leelavg reviewed (commented):
Thanks for the work, good to see the [run](https://github.com/ntkathole/pylero/actions/runs/908682545) passing as well :)

ntkathole committed:
Added 2.7 in matrix

ntkathole commented:
Thanks @leelavg for quick review. Updated!

leelavg commented:
@ntkathole ack, thank you for picking up the task, will wait for Sim to review or I'll merge it on Monday.

leelavg reviewed (approved):
LGTM

simzacks commented:
1) I can't review 165 files
2) Only .py files need to be pep8 compliant, not all the documentation. It is auto-generated, so it will go back to the way it was.
3) The previous pep8 rule was 80 characters per line (which I believe is the standard). Is there a reason you set it to 100.
I'm pretty sure that a lot of the lines that were changed to fit onto a single line were made that way because of the 80 char limit.

leelavg commented:
@simzacks allow me to answer first two
1. Review request for only ~4 files ( all dot[.] files and README.MD) and everything is a result of `black` and `generate_docs` script
2. Due to the fixer `black` code has changed and docs were regenerated i guess.

@ntkathole please address the 3rd point and revert changes done to html file which reduces the diff. Manual generation of docs is temporary and this PR doesn't need to regenerate docs.

Thanks.

simzacks commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+# configuration for pre-commit git hooks
```
I really don't love the idea of black. I started a discussion #27 , please comment there.

ntkathole replied:
I removed it for now.

simzacks commented on .flake8:
```diff
@@ -0,0 +1,2 @@
+[flake8]
+max-line-length = 100
```
why 100? It was at 80 (which I believe is the standard) and I'm pretty sure it was at least mostly pep8 compliant.

ntkathole replied:
Most of the projects started using 100 for less hassle. Thus, I thought of same. Updated to defaults, agree that we can consider it separate request.

ntkathole commented:
Thanks for review.

- Updated max-line-length to default.
- Added docs in excluded list.
- Let's consider this PR just for github action CI and pre-commit linting.
- As we want to discuss black separately, fixing all flake8 issues need separate PR, thus, keeping it disabled for now.

leelavg commented:
@ntkathole pardon causing you the trouble, I missed it noticing earlier, a lot of code is being rewritten (suppose removing unicode type checks etc, due to `pyupgrade` ig?)

This'll definitely have repercussions when run against py2, so I'll have to relook into this PR again, is that fine?

I feel, apart from fixing lint issues and rearranging imports (~isort) any other tool usage is aggressive. 

Please let me know if you've any other thoughts.

ntkathole committed:
Fixes as per comments

ntkathole commented:
Fair thought @leelavg . Good catch 👍

leelavg commented:
Taking into account that reorder_python_imports has this novel idea of reducing merge conflicts the diff is passable.

@simzacks i can merge this on your approval, we can decide on flake8 & black after discussing in #27?
- We can only introduce flake8 after deciding on fixer (black/yapf/autopep8 etc) because it'll be a difficult task to fix all pep8 issues manually.

simzacks commented:
I disagree about requiring a fixer in order to have flake8.
I'll merge this and then I'll fix the flake8 issues in the current code and then we can turn it on
